### PR TITLE
[Windows] Moves nullDevicePath to Platform header

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Moves `TestEnvironment::nullDevicePath()` to `platform.h` so it can be used in production code as well.

The reason for that is that in Windows we cant have a watcher on `NUL`. This causes integration tests to crash and it would also crash in prod as well.

Additional Description:
Tests have been verified that they are not flaky with 100+ runs without failures.

Risk Level: Low
Testing: Already tests
Docs Changes: N/A (Windows support is not released)
Release Notes: N/A (Windows support is not released)

### CODE REVIEW

File: include/envoy/common/platform.h [hunk 1]
```diff
@@ -142,6 +143,9 @@ struct msghdr {
 #define SOCKET_ERROR_INVAL WSAEINVAL
 #define SOCKET_ERROR_ADDR_IN_USE WSAEADDRINUSE
 
+namespace platform {
+constexpr std::string_view null_device_path{"NUL"};
```
Comment (wrowe) [inline include/envoy/common/platform.h:147]: absl::string_view? (fix include above)

File: include/envoy/common/platform.h [hunk 2]
```diff
@@ -227,6 +231,9 @@ typedef int os_fd_t;
 #define SOCKET_ERROR_INVAL EINVAL
 #define SOCKET_ERROR_ADDR_IN_USE EADDRINUSE
 
+namespace platform {
+constexpr std::string_view null_device_path{"/dev/null"};
```
Comment (wrowe) [inline include/envoy/common/platform.h:235]: absl::string_view?

File: include/envoy/common/platform.h [hunk 3]
```diff
@@ -142,6 +143,9 @@ struct msghdr {
 #define SOCKET_ERROR_INVAL WSAEINVAL
 #define SOCKET_ERROR_ADDR_IN_USE WSAEADDRINUSE
 
+namespace platform {
```
Comment (sunjayBhatia) [inline include/envoy/common/platform.h:146]: Even though the Google style guide mentions using lowercase namespace names, seems we have capitalized the first letter all throughout Envoy

File: include/envoy/common/platform.h (hunk 1 above)
Comment (wrowe) [reply include/envoy/common/platform.h:147]: Actually, my comment may be out of date, @lizan are we reverting from absl to std for string_view with the adoption of c++17?

**Comment by wrowe** [comment]
Note this issue in server_test at 197 and again at 216;
```
test/server/server_test.cc(197): note: Reason: cannot convert from 'const std::string_view' to 'const std::string'
test/server/server_test.cc(197): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
bazel-out/x64_windows-opt/bin/include/envoy/filesystem/_virtual_includes/filesystem_interface\envoy/filesystem/filesystem.h(93): note: see declaration of 'Envoy::Filesystem::Instance::fileExists'
```

I doubt there is a strong reason that fileExists() could not accept string_view

File: include/envoy/common/platform.h (hunk 1 above)
Comment (davinci26) [reply include/envoy/common/platform.h:147]: Based on the latest comment I opted for std::string_view

File: include/envoy/common/platform.h (hunk 3 above)
Comment (davinci26) [reply include/envoy/common/platform.h]: Thanks for pointing it out I fixed it

**Comment by davinci26** [comment]
> I doubt there is a strong reason that fileExists() could not accept string_view

I did not change the implementation because (i) all the function signatures in the `filesystem` interface take an `std::string` as input. Also (ii) the implementation of `fileExists()`relies on std::ifstream constructor. The constructor does not have an overload that accepts a`string_view` as input.

**Comment by wrowe** [comment]
> Also (ii) the implementation of `fileExists()`relies on std::ifstream constructor.

Thanks, now I recall tripping over this same puzzle before. I agree, explicitly creating temporary std::string(s) in test/server/server_test.cc should button up this commit, LGTM.

**Review by wrowe — APPROVED** [review]
Pending CI passing, this looks ready for merge.

**Comment by mattklein123** [comment]
Please fix format.

/wait

File: include/envoy/common/platform.h [hunk 4]
```diff
@@ -227,6 +231,9 @@ typedef int os_fd_t;
 #define SOCKET_ERROR_INVAL EINVAL
 #define SOCKET_ERROR_ADDR_IN_USE EADDRINUSE
 
+namespace Platform {
+constexpr std::string_view null_device_path{"/dev/null"};
```
Comment (mattklein123) [inline include/envoy/common/platform.h:235]: I thought we had a linter on this, but please use absl::string_view for now, as I think there are issues with C++17 string view on some platforms.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Small comment, thank you.

/wait

**Comment by davinci26** [comment]
what is the issue with `#include "absl/strings/string_view.h"` in `platform.h`, clang-tidy complains that the file is not found but this is the common way we include absl throughout the codebase.

I also tried to include it with `#include <absl/strings/string_view.h>` but then `CodeQL` fails.

File: include/envoy/common/platform.h (hunk 4 above)
Comment (davinci26) [reply include/envoy/common/platform.h]: thanks did not know about this. It is fixed in 1afe60a360e801678b7edca96b07ea4ce2158edd

**Comment by mattklein123** [comment]
Can you make sure the bazel build rule for the header includes a dep on string_view?

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**